### PR TITLE
Implementation of the BlockStore

### PR DESCRIPTION
**Pull request by schemar**
The block store stores headers of blocks. Not all headers have to be
reported. It is only required to report all headers that should become
justified checkpoints.
The block store tracks all justifications and finalisations through
calls to the `justify()` method. Only the polling place can call that
method.

Tests are written in one file per method that is tested.

The origin and the auxiliary block stores were deleted. The `BlockStore` is now equal to the `OriginBlockStore` and the `AuxiliaryBlockStore` must be rewritten.

### CODE REVIEW

File: test/core/block_store/justify.js [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+        blockHashAtTwentyFive = '0x1cac15d4c638c16c83562d04a27d8c977939d8700b3b5c6c71333c29fd03341a';
+        unknownBlockHash = '0xabcdef3f8dec11e46a0b636e1f85b4f4ff71161ef6111aa5092661c5008df8de'
```
Comment (deepesh-kn) [inline test/core/block_store/justify.js]: `;` is missing at the end

File: contracts/core/BlockStore.sol [hunk 2]
```diff
@@ -14,13 +14,190 @@ pragma solidity ^0.4.23;
...
+        require(
+            _initialBlockHash != bytes32(0),
+            "Initial block hash must not be zero."
+        );
+        require(
+            _initialStateRoot != bytes32(0),
+            "Initial state root must not be zero."
+        );
+        require(
+            _initialBlockHeight % _epochLength == 0,
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/core/BlockStore.sol]: `SafeMath` does not include `mod`. I will add it.

File: contracts/core/BlockStore.sol [hunk 3]
```diff
@@ -48,19 +230,394 @@ contract BlockStore is BlockStoreInterface {
...
+        (blockValid, reason) = isSourceValid(_sourceBlockHash);
+        require(blockValid, reason);
+
+        (blockValid, reason) = isTargetValid(
+            _sourceBlockHash,
+            _targetBlockHash
+        );
+        require(blockValid, reason);
+
+        isValidTransitionHash(_sourceBlockHash, _sourceBlockHash);
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/core/BlockStore.sol]: Wow 😳 Good catch.

File: contracts/core/BlockStoreInterface.sol [hunk 4]
```diff
@@ -50,8 +51,48 @@ interface BlockStoreInterface {
...
      */
-    function getLatestBlockHeight()
+    function latestBlockHeight()
         external
         view
         returns (uint256 height_);
+
+    /**
+     * @notice Validates a given vote. For a vote to be valid:
+     *         - The transition object must be correct and
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/core/BlockStoreInterface.sol]: We could. Would you prefer that? I thought it reads more natural like this (more like a sentence).

**Comment by schemar** [comment]
@deepesh-kn I decided to use named parameters everywhere. It does not increase the gas cost of execution. And in my opinion it makes the code more readable as it clearly states what it returns. Kind of like
```solidity
returns (boos isValid_)
```
instead of
```solidity
returns (bool /* isValid */)
```

I can remove it if we think it makes code harder to read or maintain. However we should agree on something and be consistent. What do you think?

/cc @sarvesh-ost

**Comment by schemar** [comment]
We decided to remove the named return variables where they are not needed. There will generally be cases in solidity where return variables cannot be named due to the limit in local variables per function.

**Pull request merged**